README.md
Andfun Yangon/Studio
This is auto capturing images from multiple cameras.

Version: 2.0


Objective: 
    Multiple Web cams capturing imgaes per second automatically.


Description:
    A user can take images automatically from multiple cameras connected to his computer. User can define image width, height and photo as much as he wants. 


File Description:

    "usbVideoDevice.py":
        a class to detect all connected wbcam names, webcam_ids.

    "auto_capture_config.py":
        Only if we run auto_capture_config.py, webcam devices connected to a computer will be got.
        No porblem whatever webcam ids are. 
        This code will read connected webcam ids from "UsbVideoDevice class" and write config data to "auto_capture_config.txt".

    "auto_capture_config.txt":
         input settings that will satisfy user's requirement

    "auto_capture.py":
         will do the main objective.



Prerequisite:

    Linux environment, python3, opencv-python
    sudo apt install python3, python3-pip, python3-tk
    pip3 install opencv-python
    

How to run:

    python auto_capture_config.py
    python auto_capture.py


notes/release-notes
Andfun Yangon/Studio
This is auto capturing images from multiple cameras.

Version: 2.0


New Features: 
    Multiple Web cam capturing 3 imgaes per second automatically.


Description:
    I set the maximum limit of auto capturing imgaes just to 20 phtots.


File Description:
    "usbVideoDevice.py":
        a class to know all connected  wbcam names, webcam_ids and  ports.

    "auto_capture_config.py":
        Only if we run auto_capture_config.py, we will get user's connect web cam ids. 
        No porblem whatever webcam ids are. 
        This code will read connected webcam ids from "UsbVideoDevice class" and write config data to "auto_capture_config.txt".
        In this file, set "save_img_path" where you want to store images. Create folder and set path to it.


    "auto_capture_config.txt":
         input settings that will satisfy user's requirement

    "auto_capture.py":
         will do the main function.


Prerequisite:
    Linux environment, python3, opencv-python
    sudo apt install python3, python3-pip, python3-tk
    pip3 install opencv-python
    

How to run:
    python auto_capture_config.py
    python auto_capture.py



test.py
ls = [(0, 1, 2), ['a', 'b']]
dic = {'a': [0, 1, 2], 'b': [4, 5]}

ls_dic = [
    {'a': [0, 1, 2], 'b':[4, 5]},
    {'id': [10, 20], 'name':['aye', 'bo']},
    (1, 2, 3, 4)
]

print(ls[0][1])
print(dic['a'][2])
print(ls_dic[1]['name'][0])


test/config.py
import configparser

config = configparser.ConfigParser()

config['DEFAULT'] = {
    'ServerAliveInterval': '45',
    'Compression': 'yes',
    'CompressionLevel': '9'
}

config['bitbucket.org'] = {}
config['bitbucket.org']['User'] = 'hg'

config['topsecret.server.com'] = {}

topsecret = config['topsecret.server.com']
topsecret['Port'] = '50022'     # mutates the parser
topsecret['ForwardX11'] = 'no'  # same here

config['DEFAULT']['ForwardX11'] = 'yes'

config['content'] = {}
config['content']['cost'] = "[20, 10.5, -1]"


with open('config.ini', 'w') as configfile:
    config.write(configfile)


test/config_reader.py
import configparser
import json

config = configparser.ConfigParser()

config.sections()   # []
config.read('config.ini')  # ['config.ini']
# config.read('/home/wc/Desktop/code/OpenCV/AndFun/config.ini')
config.sections()   # ['bitbucket.org', 'topsecret.server.com']

'bitbucket.org' in config   # True
'bytebong.com' in config    # False

config['bitbucket.org']['User']     # 'hg'
config['DEFAULT']['Compression']    # 'yes'

topsecret = config['topsecret.server.com']
topsecret['ForwardX11']     # 'no'
topsecret['Port']   # '50022'

# for key in config['topsecret.server.com']:
    # print(key)  # user
    # compressionlevel
    # serveraliveinterval
    # compression
    # forwardx11

config['bitbucket.org']['ForwardX11']   # yes

config.items('content')
# [
#     ('serveraliveinterval', '45'), ('compression', 'yes'), ('compressionlevel', '9'),
#     ('forwardx11', 'yes'), ('cost', '[20, 10.5, -1]')
# ]
dict(config.items('content'))
# {
#     'serveraliveinterval': '45', 'compression': 'yes', 'compressionlevel': '9', 
#     'forwardx11': 'yes', 'cost': '[20, 10.5, -1]'
# }
json.loads(config.get("content", "cost"))
# [20, 10.5, -1]    # list


usbVideoDevice.py
import subprocess

class UsbVideoDevice:
    def __init__(self):

        self.webcams = {'ids': [], 'names': [], }

        try:
            cmd = 'ls -la /dev/v4l/by-id'
            res = subprocess.check_output(cmd.split())
            by_id = res.decode()
        except Exception:
            return

        # Set connected webcam device names and ids
        for line in by_id.split('\n'):
            if('../../video' in line):
                tmp = line.split('usb-')
                tmp = tmp[1].split('-video-index')

                if tmp[0] not in self.webcams['names']:
                    self.webcams['names'].append(tmp[0])

                tmp = tmp[1].split('../../video')

                try:
                    # if tmp[1] not in self.webcams['ids'] and int(tmp[1]) % 2 == 0:
                    self.webcams['ids'].append(int(tmp[1]))
                except ValueError:
                    print("usbVideoDevice.py can't change to int.")

        print("Usb Video Devices:\n", self.webcams)
        print('\n')

    # Get connected webcam ids to take images
    def getAllIds(self):
        return self.webcams['ids']

    def getAllNames(self):
        return self.webcams['names']


auto_capture_bk.py
import cv2
import os
import datetime


start_time = datetime.datetime.now()
print(f'{start_time.hour}hr {start_time.minute}min {start_time.second}sec' )
img_counter = 1

# cap = cv2.VideoCapture('/dev/video0')
cap = cv2.VideoCapture(2)
cap.set(3, 1920)
cap.set(4, 1080)    # to get FHD, but my display is HD(1280x720)

while True:
    ret, img = cap.read()
    cv2.imshow('Studio', img)

    if (datetime.datetime.now() - start_time).seconds == 1: # Time elapsed 1 sec
        start_time = datetime.datetime.now()
        print(start_time)

        freq = 0
        while freq < 3:
            # img_name = "prefix_camera{}.png".format(img_counter)
            img_name = f"prefix_camera{img_counter}.png"

            path = os.path.expanduser("/usr/aaa") 
            # sudo chmod 777 /usr/aaa in terminal, making executable to solve write permission denied

            cv2.imwrite(os.path.join(path, img_name), img)
            print(f"{img_name} written! {img.shape[1]}x{img.shape[0]} pixels")

            img_counter += 1
            freq += 1

    if img_counter == 31:   break

    if cv2.waitKey(10) & 0xFF == ord('q'):
        break

cap.release()  # close the camera
cv2.destroyAllWindows()  # close all the opened windows


auto_capture_config.py
import configparser
import tkinter as tk
# from tkinter import *
from tkinter import filedialog, messagebox
from usbVideoDevice import UsbVideoDevice


class Window(tk.Tk):
    def __init__(self):
        super().__init__()
        self.title("Webcam Config Settings")

        tk.Label(self, text="Imgae Width:").grid(row=0, column=0)
        tk.Label(self, text="Image Height:").grid(row=1, column=0)
        tk.Label(self, text="Captured Image per second:").grid(row=2, column=0)
        tk.Label(self, text="Max Image to capture:").grid(row=3, column=0)
        tk.Label(self, text="Saved Image Path:").grid(row=4, column=0)

        is_digit = self.register(self.only_numbers)
        self.img_width = tk.Entry(
            self, validate='key', validatecommand=(is_digit, '%S'))
        self.img_width.grid(row=0, column=1, sticky=tk.W, padx=20, pady=20)

        self.img_height = tk.Entry(
            self, validate='key', validatecommand=(is_digit, '%S'))
        self.img_height.grid(row=1, column=1, sticky=tk.W, padx=20, pady=20)

        self.capture_img_per_sec = tk.Entry(
            self, validate='key', validatecommand=(is_digit, '%S'))
        self.capture_img_per_sec.grid(
            row=2, column=1, sticky=tk.W, padx=20, pady=20)

        self.max_img_counter = tk.Entry(self)
        self.max_img_counter.config(
            validate='key', validatecommand=(is_digit, '%S'))
        self.max_img_counter.grid(
            row=3, column=1, sticky=tk.W, padx=20, pady=20)

        self.dir_txt = tk.StringVar()
        self.save_img_path = tk.Button(
            self, text='Browse', command=self.browse)
        self.save_img_path.grid(row=4, column=1, sticky=tk.W, padx=20, pady=20)

        if self.dir_txt:
            tk.Label(self, textvar=self.dir_txt,
                     fg='blue').grid(row=5, column=1)

        tk.Button(self, text="Cancel", command=self.destroy).grid(
            row=6, column=1, sticky=tk.W, padx=20, pady=20)
        tk.Button(self, text="Confirm", bg='blue', fg='white', command=self.confirm).grid(
            row=6, column=1, sticky=tk.E, padx=20, pady=20)

    def browse(self):
        self.dir_selected = filedialog.askdirectory()
        self.dir_txt.set(self.dir_selected)
        print(self.dir_txt.get())

    def only_numbers(self, char):
        return char.isdigit()

    def confirm(self):
        # if False:     # Test
        #     return

        if not self.img_width.get() or not self.img_height.get() or not self.capture_img_per_sec.get() or not self.max_img_counter.get():
            messagebox.showerror("Error", "Entry fields cannot be empty.")

        elif not self.dir_txt.get():
            messagebox.showwarning('Error', "Saved image path is required.")

        else:
            self.capture_img_per_sec = self.capture_img_per_sec.get() if int(self.capture_img_per_sec.get()) in range(1, 11) else 3

            config['content'] = {
                # get() returns str
                'webcam_id': webcam_ids,  # [0, 2]
                'webcam_names': webcam_names,
                'img_width': self.img_width.get(),  # 1280
                'img_height': self.img_height.get(),  # 720
                'capture_img_per_sec': self.capture_img_per_sec,  # 3
                'max_img_counter': self. max_img_counter.get(),   # 20
                'save_img_path': self.dir_txt.get()
                # 'save_img_path': '/home/wc/Desktop/code/ML/andfun/images'
            }

            self.destroy()

            with open('auto_capture_config.txt', 'w') as configfile:
                config.write(configfile)

            print("Config data is saved to auto_capture_config.txt")


usbVideoDevice = UsbVideoDevice()
webcam_ids = usbVideoDevice.getAllIds()  # [0, 2]
webcam_names = usbVideoDevice.getAllNames()

config = configparser.ConfigParser()

window = Window()
window.mainloop()


auto_capture.py
import cv2
import os
import datetime
import time
import configparser
import json

# Read user configuration data
config = configparser.ConfigParser()
config.read('auto_capture_config.txt')

# webcam_names = json.loads(config.get('content', 'webcam_names'))
# print(webcam_names)
webcam_ids = json.loads(config.get('content', 'webcam_id'))

content = config['content']
img_width = int(content['img_width'])
img_height = int(content['img_height'])
config_path = os.path.expanduser(content['save_img_path'])

# Reading multiple web cams
cap, cap_titles = ([None] * len(webcam_ids) for _ in range(2))
unopened_webcam_ids = []

for index, webcam_id in enumerate(webcam_ids):
    if cv2.VideoCapture(webcam_id).isOpened():
        cap[index] = cv2.VideoCapture(webcam_id)    # 0, 2
        # cap[index].set(3, 640)   # 1280
        # cap[index].set(4, 480)  # 720
        cap_titles[index] = 'Webcam: ' + str(webcam_id)
    else:
        unopened_webcam_ids.append(webcam_id)
        print(f"Failed to open webcam id:{webcam_id}")


# Cleaning failed index of which webcam ids cannot be read
webcam_ids = [ele for ele in webcam_ids if ele not in unopened_webcam_ids]
cap_titles = [ele for ele in cap_titles if ele is not None]

for ele in cap:
    if ele is None:
        cap.remove(ele)


# Initialize fixed size lists
ret, imgs = [list(range(len(webcam_ids))) for _ in range(2)]

# Separating dir using webcam_ids
paths = [None] * len(webcam_ids)
for i in range(len(webcam_ids)):
    paths[i] = os.path.join(config_path, 'webcam' + str(webcam_ids[i]))

    if not os.path.exists(paths[i]):    # '/usr/aaa'
        try:
            os.makedirs(paths[i])
        except OSError:
            print("Can't create destination directory (%s)!" % paths[i])
            home = os.path.expanduser("~")
            path = os.path.join(home, 'unkonw_path')
            # path = os.path.join(home, 'aaa')
            os.makedirs(path, exist_ok=True)
            print(f"Default path will be used: {path}")

# Image Capturing process
start_time = datetime.datetime.now()
img_counter = 0
while True:

    for i in range(len(webcam_ids)):
        # if cap[i] is None:  continue
        ret[i], imgs[i] = cap[i].read()
        cv2.imshow(cap_titles[i], imgs[i])

    # time.sleep(1)
    if (datetime.datetime.now() - start_time).seconds == 1:  # Time elapsed 1 sec
        start_time = datetime.datetime.now()
        txt_on_imgs = start_time.strftime("%Y_%m_%d_%Hhr_%Mmin_%Ssec")

        print("Writing Images . . .")
        # print(start_time.second)

        for _ in range(int(content['capture_img_per_sec'])):
            img_counter += 1
            for j in range(len(webcam_ids)):

                # if isinstance(imgs[j], int):    continue

                img_name = f"{txt_on_imgs}_{img_counter}.png"

                imgs[j] = cv2.resize(imgs[j], (img_width, img_height))
                cv2.putText(imgs[j], f"id:{webcam_ids[j]}", (int(img_width * 0.6), int(img_height * 0.9)), cv2.FONT_HERSHEY_COMPLEX, 1, (0, 0, 0), 2)

                isWritten = cv2.imwrite(
                    os.path.join(paths[j], img_name), imgs[j])
                # print(f"{img_name} written! {imgs[j].shape[1]}x{imgs[j].shape[0]} pixels")


    if img_counter >= int(content['max_img_counter']):
        break

    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

cv2.destroyAllWindows()  # close all the opened windows

try:
    if isWritten:
        print("Saved Image Path:" + config_path)
except NameError:
    print("Images are not saved.")
